### PR TITLE
Joint Polarities

### PR DESCRIPTION
**Pull request by MaximilienNaveau**
## What Changed

- Add a setter in BlmcJointModule in order to change the joint polarity.
- Add a setter in BlmcJointModule**s** in orde to change all joints polarities at once.
- Update the Solo robot with the good polarities.

## API update

- This PR intentionally did not modify the API of the BlmcJointModule**s** as robots are shared among several lab
- Another PR may involve the update of all blmc_robots to a New API including the joint polarity in the constructor and the set_motor_array?

## How did I tested it.

- I **will** test it on the robot tomorrow.

### CODE REVIEW

**Comment by jviereck** [comment]
Is the word "polarities" used? I would have used "direction" (just not sure which one is more common used in robotics and in the English language).

File: include/blmc_robots/blmc_joint_module.hpp [hunk 1]
```diff
@@ -118,6 +118,13 @@ class BlmcJointModule
      */
     void set_zero_angle(const double& zero_angle);
 
+    /**
+     * @brief reverse the joint polarity rotation axis
+     * 
+     * @param reverse_polarity true:reverse rotation axis, false:do nothing
+     */
+    void reverse_joint_polarity(const bool& reverse_polarity);
```
Comment (jviereck) [inline include/blmc_robots/blmc_joint_module.hpp]: call this `toggle_joint_polartiy`? "Toggle" is what is used in software usually.

File: include/blmc_robots/blmc_joint_module.hpp [hunk 2]
```diff
@@ -401,10 +408,22 @@ class BlmcJointModules
...
 
+    /**
+     * @brief Set the polarities of the joints
+     * 
+     * @param polarities 
+     */
+    void reverse_joint_polarities(std::array<bool, COUNT> polarities)
+    {
+        for(size_t i = 0; i < COUNT; i++)
+        {
```
Comment (jviereck) [inline include/blmc_robots/blmc_joint_module.hpp]: What was the agreement we found during the software discussion meeting? Should it be like this or `for (...) {`?

File: include/blmc_robots/blmc_joint_module.hpp [hunk 3]
```diff
@@ -401,10 +408,22 @@ class BlmcJointModules
...
+            modules_[i]->reverse_joint_polarity(polarities[i]);
+        }
+    }
     /**
      * @brief Register the joint torques to be sent for all modules.
      * 
-     * @param desired_torques 
+     * @param desired_torques
```
Comment (jviereck) [inline include/blmc_robots/blmc_joint_module.hpp]: Please provide unites.

File: include/blmc_robots/solo.hpp [hunk 4]
```diff
@@ -334,7 +334,7 @@ class Solo
   /**
    * @brief Address the rotation direction of the motor
```
Comment (jviereck) [inline include/blmc_robots/solo.hpp]: Please have "." at end of brief sentences.

File: src/blmc_joint_module.cpp [hunk 5]
```diff
@@ -56,6 +56,10 @@ void BlmcJointModule::set_zero_angle(const double& zero_angle)
     zero_angle_ = zero_angle;
 }
 
+void BlmcJointModule::reverse_joint_polarity(const bool& reverse_polarity)
+{
+    polarity_ = reverse_polarity ? -1.0:1.0;
```
Comment (jviereck) [inline src/blmc_joint_module.cpp]: Missing space `reverse_polarity ? -1.0[ ]:[ ]1.0;`

File: include/blmc_robots/blmc_joint_module.hpp (hunk 1 above)
Comment (jviereck) [reply include/blmc_robots/blmc_joint_module.hpp]: Should there also be a `set_joint_polarity(const bool& should_reverse)`?

**Comment by jviereck** [comment]
Added some comments - some are general fly by comments not related to these change. Overall, I think this looks good (though I am not super familiar with the code, so I don't approve).

Leaving final approval to @mwuethri as he is more familiar with the codebase.

Overall, thanks for working on this Max!

File: include/blmc_robots/blmc_joint_module.hpp (hunk 2 above)
Comment (MaximilienNaveau) [reply include/blmc_robots/blmc_joint_module.hpp]: We said that brackets should always comes next line.

File: include/blmc_robots/blmc_joint_module.hpp (hunk 3 above)
Comment (MaximilienNaveau) [reply include/blmc_robots/blmc_joint_module.hpp]: I will do this, though this is not really the objective of this (I guess I do not like trailing space)

File: include/blmc_robots/blmc_joint_module.hpp (hunk 1 above)
Comment (MaximilienNaveau) [reply include/blmc_robots/blmc_joint_module.hpp]: I actually like the `set_joint_polarity(const bool& should_reverse)` I will to that in both classes

**Comment by MaximilienNaveau** [comment]
Thanks Julian for the feedback, it is helpful!

**Comment by MaximilienNaveau** [comment]
@mwuethri Wait that I take into account Julian's comment before your review ;)

**Comment by mwuethri** [comment]
> @mwuethri Wait that I take into account Julian's comment before your review ;)

@MaximilienNaveau just let me know whenever i should review

**Comment by MaximilienNaveau** [comment]
@mwuethri you can review now.

File: include/blmc_robots/blmc_joint_module.hpp [hunk 6]
```diff
@@ -106,18 +106,25 @@ class BlmcJointModule
     /**
      * @brief Set the joint torque to be sent.
      * 
-     * @param desired_torque 
+     * @param desired_torque (N/m)
```
Comment (mwuethri) [inline include/blmc_robots/blmc_joint_module.hpp]: this is not N/m but Nm right?

File: include/blmc_robots/blmc_joint_module.hpp [hunk 7]
```diff
@@ -106,18 +106,25 @@ class BlmcJointModule
...
+     * @param zero_angle (rad)
      */
     void set_zero_angle(const double& zero_angle);
 
+    /**
+     * @brief Define if the motor should turn clock-wize or counter clock-wize.
+     * 
+     * @param should_reverse true:reverse rotation axis, false:do nothing.
+     */
+    void set_joint_polarity(const bool& should_reverse);
```
Comment (mwuethri) [inline include/blmc_robots/blmc_joint_module.hpp]: should_reverse seems not very descriptive to me, could we call it reverse_polarity?

File: include/blmc_robots/blmc_joint_module.hpp [hunk 8]
```diff
@@ -127,74 +134,75 @@ class BlmcJointModule
     /**
      * @brief Get the sent joint torque.
      * 
-     * @return double 
+     * @return double (N/m).
```
Comment (mwuethri) [inline include/blmc_robots/blmc_joint_module.hpp]: Nm

File: include/blmc_robots/blmc_joint_module.hpp [hunk 9]
```diff
@@ -127,74 +134,75 @@ class BlmcJointModule
...
      */
     double get_sent_torque() const;
 
     /**
      * @brief Get the measured joint torque.
      * 
-     * @return double 
+     * @return double (N/m).
```
Comment (mwuethri) [inline include/blmc_robots/blmc_joint_module.hpp]: Nm

File: include/blmc_robots/blmc_joint_module.hpp [hunk 10]
```diff
@@ -127,74 +134,75 @@ class BlmcJointModule
...
+     * @return double (rad).
      */
     double get_zero_angle() const;
 
     /**
      * @brief Set control gains for PD position controller.
      *
-     * @param kp P gain.
-     * @param kd D gain.
+     * @param kp P gain ( (N/m) / rad ).
```
Comment (mwuethri) [inline include/blmc_robots/blmc_joint_module.hpp]: Nm

File: include/blmc_robots/blmc_joint_module.hpp [hunk 11]
```diff
@@ -127,74 +134,75 @@ class BlmcJointModule
...
+     * @param kd D gain ( (N/m) / (rad/s) ).
```
Comment (mwuethri) [inline include/blmc_robots/blmc_joint_module.hpp]: Nm

File: include/blmc_robots/blmc_joint_module.hpp [hunk 12]
```diff
@@ -127,74 +134,75 @@ class BlmcJointModule
...
     void set_position_control_gains(double kp, double kd);
 
     /**
      * @brief Execute one iteration of the position controller.
      *
-     * @param target_position_rad  Target position.
+     * @param target_position_rad  Target position (rad).
      *
-     * @return Torque command.
+     * @return Torque command (N/m).
```
Comment (mwuethri) [inline include/blmc_robots/blmc_joint_module.hpp]: Nm

File: include/blmc_robots/blmc_joint_module.hpp [hunk 13]
```diff
@@ -401,10 +408,23 @@ class BlmcJointModules
         }
     }
 
+    /**
+     * @brief Set the polarities of the joints
+     * (see BlmcJointModule::set_joint_polarity)
+     * 
+     * @param should_reverse
+     */
+    void set_joint_polarities(std::array<bool, COUNT> should_reverse)
```
Comment (mwuethri) [inline include/blmc_robots/blmc_joint_module.hpp]: as above, should_reverse does not seem like a very clear name to me, could we call it reverse_polarities?

File: src/blmc_joint_module.cpp [hunk 14]
```diff
@@ -56,6 +56,10 @@ void BlmcJointModule::set_zero_angle(const double& zero_angle)
     zero_angle_ = zero_angle;
 }
 
+void BlmcJointModule::set_joint_polarity(const bool& should_reverse)
+{
+    polarity_ = should_reverse ? -1.0 : 1.0;
```
Comment (mwuethri) [inline src/blmc_joint_module.cpp]: should_reverse -> reverse_polarity

File: include/blmc_robots/blmc_joint_module.hpp (hunk 6 above)
Comment (MaximilienNaveau) [reply include/blmc_robots/blmc_joint_module.hpp]: Fair enough I will correct this.

**Comment by MaximilienNaveau** [comment]
Done

**Comment by jviereck** [comment]
Post merge comment: The proper way to notate unites is using `[unit]`, so the comments should be reformatted `zero_angle (rad)` -> `zero_angle [rad]`.

**Pull request merged**